File: interview_cake/stacks_and_queues/parentheses.rb
def get_closing_paren(sentence, index)

  # Find the position of the matching closing parenthesis.
  if sentence[index] != '('
    raise ArgumentError, "Index must point to an opening parenthese."
  end
  
  open_count = 0
  
  (index + 1).upto(sentence.length - 1) do |i|
    char = sentence[i]
    if char == '('
      open_count += 1
    elsif char == ')' && open_count == 0
      return i
    elsif char == ')'
      open_count -= 1
    end
  end

  raise StandardError, "There is no closing parenthese!"
end


















# Tests

def run_tests
  desc = 'all openers then closers'
  actual = get_closing_paren('((((()))))', 2)
  expected = 7
  assert_equal(actual, expected, desc)

  desc = 'mixed openers and closers'
  actual = get_closing_paren('()()((()()))', 5)
  expected = 10
  assert_equal(actual, expected, desc)

  desc = 'no matching closer'
  assert_raises(desc) { get_closing_paren('()(()', 2) }
end

def assert_equal(a, b, desc)
  puts "#{desc} ... #{a == b ? 'PASS' : "FAIL: #{a.inspect} != #{b.inspect}"}"
end

def assert_raises(desc)
  yield
  puts "#{desc} ... FAIL"
rescue
  puts "#{desc} ... PASS"
end

run_tests

File: interview_cake/stacks_and_queues/parentheses.js
function getClosingParen(sentence, index) {

  // Find the position of the matching closing parenthesis
  if (sentence[index] !== '(') {
    throw "The index must point to an opening parenthese!"
  } 
  
  let openCount = 0;
  
  for (let i = index + 1; i < sentence.length; i++) {
    let char = sentence[i];
    
    if (char == '(') {
      openCount += 1;
    } else if (char == ')' && openCount == 0) {
      return i;
    } else if (char == ')') {
      openCount -= 1;
    }
  }

  throw "There is no closing parenthese!"
}


















// Tests

let desc = 'all openers then closers';
let actual = getClosingParen('((((()))))', 2);
let expected = 7;
assertEqual(actual, expected, desc);

desc = 'mixed openers and closers';
actual = getClosingParen('()()((()()))', 5);
expected = 10;
assertEqual(actual, expected, desc);

desc = 'no matching closer';
const noCloser = () => (getClosingParen('()(()', 2));
assertThrowsError(noCloser, desc);

function assertEqual(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`);
  }
}

function assertThrowsError(func, desc) {
  try {
    func();
    console.log(`${desc} ... FAIL`);
  } catch (e) {
    console.log(`${desc} ... PASS`);
  }
}
